File: 300-longest-increasing-subsequence/300-longest-increasing-subsequence.java
class Solution {
    public int lengthOfLIS(int[] nums) {
        int n = nums.length;
        int[] dp = new int[n];
        int res = 0;
        for(int i = 0; i < n; i++) {
            int idx = Arrays.binarySearch(dp, 0, res, nums[i]);
            if(idx < 0) idx = -idx - 1;
            dp[idx] = nums[i];
            if(idx == res) res++;
        }
        return res;
    }
}

File: 300-longest-increasing-subsequence/300-longest-increasing-subsequence.py
import bisect
class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        sub, n = [], len(nums)
        for i in range(n):
            j = bisect.bisect_left(sub, nums[i])
            if j == len(sub):
                sub.append(nums[i])
            else:
                sub[j] = nums[i]
        return len(sub)
                